0000-navigator-view-api.md

```diff
@@ -0,0 +1,93 @@
+* Start Date: 2018-01-29
+* RFC PR: (leave this empty)
+* React Navigation Issue: (leave this empty)
+
+# Summary
+
+Introduce a simpler API for createNavigator and navigation views. By providing "sceneDescriptors" to the navigation view, it does not need to compute options or manually create the child navigation prop.
+
+Down the line, this will enable us to deprecate `addNavigationHelpers`, which is an avoidable burden for redux apps and when implementing custom navigators. Eventually, each navigator may provide a different set of helpers.
+
+# Basic example
+
+Using a new `createNavigator` API and `screenDescriptors` prop, this is simplest possible custom navigation view:
+
+```js
+const CustomNavigationView = ({navigation, screenDescriptors}) => {
+  const {routes, index}
+  const descriptor = screenDescriptors.find(d => d.key === routes[index].key);
+  const ScreenComponent = descriptor.getComponent();
+  return (
+    <ScreenComponent
+      navigation={descriptor.navigation}
+      options={descriptor.options}
+    />
+  );
+};
+const router = StackRouter({...}, {...});
+const CustomNavigator = createNavigator(CustomNavigationView, router);
+```
+
+This `<CustomNavigator />` will have behave like a stack, and will be able to present screens registered within `router`.
+
+# Motivation
+
+This change prevents the router from being passed directly into the navigation view, which was causing coupling between the concepts. This change will allow for faster innovation on the navigation views, and allow them to be consumed by other navigation libraries.
+
+This also makes it easier to implement navigation views, because the children navigation props have already been provided within `props.sceneDescriptors`
+
+Why are we doing this? What use cases does it support? What is the expected outcome?
+
+Please focus on explaining the motivation so that if this RFC is not accepted, the motivation could be used to develop alternative solutions. In other words, enumerate the constraints you are trying to solve without coupling them too closely to the solution you have in mind.
+
+# Detailed design
+
+Navigators will now be created via a simpler API:
+
+`const Navigator = createNavigator(NavigationView, router, navigationConfig);`
+
+In this example, navigationConfig is an object that will be passed as a prop to the navigtion view. The `NavigationView` takes the following props:
+
+* `navigation` - Navigation prop with `dispatch`, `state`, and `addListener`
+* `screenProps` - The navigator will pass this through, so that the navigation view can provide screenProps to each screen
+* `navigationConfig` - Arbitrary config object from `createNavigator`
+* `sceneDescriptors` - Array of objects, one for each `navigation.state.routes`. Each scene descriptor has:
```

Comment:
it makes me a little uneasy how similar `scene` and `screen` are. I don't have a better alternative but something to consider. what makes this a scene and not a screen? or a route, for that matter?

Reply:
yeah, I'm thinking we should just rename everything to one thing. Way back in the day we had chosen "screen". Although personally I always preferred scene over screen, because screen generally refers to the physical device screen.

In this case I had chosen scene because Transitioner has an internal scene concept which lines up with what we are describing here.